Source/Core/Core/IOS/USB/Emulated/Microphone.cpp

```diff
@@ -0,0 +1,383 @@
+  if (Config::Get(Config::MAIN_WII_SPEAK_MUTED))
+    return nframes;
+
+  auto* mic = static_cast<Microphone*>(user_data);
+  return mic->DataCallback(static_cast<const SampleType*>(input_buffer), nframes);
+}
+
+long Microphone::DataCallback(const SampleType* input_buffer, long nframes)
+{
+  std::lock_guard lock(m_ring_lock);
+
+  // Skip data if sampling is off or mute is on
+  if (!m_sampler.sample_on || m_sampler.mute)
+    return nframes;
+
+  std::span<const SampleType> buffer(input_buffer, nframes);
+  const auto gain = ComputeGain(Config::Get(Config::MAIN_WII_SPEAK_VOLUME_MODIFIER));
+  const auto apply_gain = [gain](SampleType sample) {
+    return MathUtil::SaturatingCast<SampleType>(sample * gain);
+  };
+
+  for (const SampleType le_sample : std::ranges::transform_view(buffer, apply_gain))
+  {
+    UpdateLoudness(le_sample);
+    m_stream_buffer[m_stream_wpos] = Common::swap16(le_sample);
+    m_stream_wpos = (m_stream_wpos + 1) % STREAM_SIZE;
+  }
+
+  m_samples_avail += nframes;
+  if (m_samples_avail > STREAM_SIZE)
+  {
+    WARN_LOG_FMT(IOS_USB, "Wii Speak ring buffer is full, data will be lost!");
+    m_samples_avail = STREAM_SIZE;
+  }
+
+  return nframes;
+}
+#endif
+
+u16 Microphone::ReadIntoBuffer(u8* ptr, u32 size)
+{
+  static constexpr u32 SINGLE_READ_SIZE = BUFF_SIZE_SAMPLES * sizeof(SampleType);
+
+  // Avoid buffer overflow during memcpy
+  static_assert((STREAM_SIZE % BUFF_SIZE_SAMPLES) == 0,
+                "The STREAM_SIZE isn't a multiple of BUFF_SIZE_SAMPLES");
+
+  std::lock_guard lock(m_ring_lock);
+
+  u8* begin = ptr;
+  for (u8* end = begin + size; ptr < end; ptr += SINGLE_READ_SIZE, size -= SINGLE_READ_SIZE)
+  {
+    if (size < SINGLE_READ_SIZE || m_samples_avail < BUFF_SIZE_SAMPLES)
+      break;
+
+    SampleType* last_buffer = &m_stream_buffer[m_stream_rpos];
+    std::memcpy(ptr, last_buffer, SINGLE_READ_SIZE);
+
+    m_samples_avail -= BUFF_SIZE_SAMPLES;
+    m_stream_rpos += BUFF_SIZE_SAMPLES;
+    m_stream_rpos %= STREAM_SIZE;
+  }
+  return static_cast<u16>(ptr - begin);
+}
+
+u16 Microphone::GetLoudnessLevel() const
+{
+  if (m_sampler.mute || Config::Get(Config::MAIN_WII_SPEAK_MUTED))
+    return 0;
+  return m_loudness_level;
+}
+
+// Based on graphical cues on Monster Hunter 3, the level seems properly displayed with values
+// between 0 and 0x3a00.
+//
+// TODO: Proper hardware testing, documentation, formulas...
+void Microphone::UpdateLoudness(const SampleType sample)
+{
+  // Based on MH3 graphical cues, let's use a 0x4000 window
+  static const u32 WINDOW = 0x4000;
+  static const float UNIT = (m_loudness.DB_MAX - m_loudness.DB_MIN) / WINDOW;
+
+  m_loudness.Update(sample);
+
+  if (m_loudness.samples_count >= m_loudness.SAMPLES_NEEDED)
+  {
+    const float amp_db = m_loudness.GetAmplitudeDb();
+    m_loudness_level = static_cast<u16>((amp_db - m_loudness.DB_MIN) / UNIT);
+
+#ifdef WII_SPEAK_LOG_STATS
+    m_loudness.LogStats();
+#else
+    DEBUG_LOG_FMT(IOS_USB,
```

Comment:
I know it's a debug log, but one every 128 samples is maybe excessive? Maybe just rely on the `WII_SPEAK_LOG_STATS`macro  if someone wants this logged?

Reply:
That's a valid concern and can bother someone checking other IOS_USB debug logs so I'll rely on the macro as you suggested.